domain/model/state/modelstate.go

```diff
@@ -105,6 +111,402 @@ func (s *ModelState) Delete(ctx context.Context, uuid coremodel.UUID) error {
+				return errors.Errorf("upserting constraint tags for constraint %q: %w", existingConstraint.UUID, err)
+			}
+		}
+
+		if consValue.Spaces != nil {
+			err = s.upsertContraintSpaces(ctx, tx, existingConstraint.UUID, *consValue.Spaces)
+			if err != nil {
+				return errors.Errorf("upserting constraint spaces for constraint %q: %w", existingConstraint.UUID, err)
+			}
+		}
+
+		if consValue.Zones != nil {
+			err = s.upsertContraintZones(ctx, tx, existingConstraint.UUID, *consValue.Zones)
+			if err != nil {
+				return errors.Errorf("upserting constraint zones for constraint %q: %w", existingConstraint.UUID, err)
+			}
+		}
+		return nil
+	})
+	return errors.Capture(err)
+}
+
+func (s *ModelState) upsertContraintTags(ctx context.Context, tx *sqlair.TX, constraintUUID string, tags []string) error {
+	removeConstraintTagsStmt, err := s.Prepare(`
+DELETE FROM constraint_tag
+WHERE constraint_uuid = $dbConstraintTag.constraint_uuid
+`, dbConstraintTag{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	insertConstraintTagStmt, err := s.Prepare(`
+INSERT INTO constraint_tag (*)
+VALUES ($dbConstraintTag.*)`, dbConstraintTag{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, removeConstraintTagsStmt, dbConstraintTag{ConstraintUUID: constraintUUID}).Run()
+	if err != nil {
+		return errors.Errorf("removing existing constraint tags for constraint %q: %w", constraintUUID, err)
+	}
+
+	if len(tags) == 0 {
+		return nil
+	}
+
+	for _, tag := range tags {
+		if tag == "" {
+			continue
+		}
+		err := tx.Query(ctx, insertConstraintTagStmt, dbConstraintTag{
+			ConstraintUUID: constraintUUID,
+			Tag:            tag,
+		}).Run()
+		if err != nil {
+			return errors.Errorf("inserting constraint tag %q: %w", tag, err)
+		}
+	}
+	return nil
+}
+
+func (s *ModelState) upsertContraintSpaces(ctx context.Context, tx *sqlair.TX, constraintUUID string, spaces []string) error {
+	removeConstraintSpacesStmt, err := s.Prepare(`
+DELETE FROM constraint_space
+WHERE constraint_uuid = $dbConstraintSpace.constraint_uuid
+`, dbConstraintSpace{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	insertConstraintSpaceStmt, err := s.Prepare(`
+INSERT INTO constraint_space (*)
+VALUES ($dbConstraintSpace.*)`, dbConstraintSpace{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, removeConstraintSpacesStmt, dbConstraintSpace{ConstraintUUID: constraintUUID}).Run()
+	if err != nil {
+		return errors.Errorf("removing existing constraint spaces for constraint %q: %w", constraintUUID, err)
+	}
+
+	if len(spaces) == 0 {
+		return nil
+	}
+
+	for _, space := range spaces {
+		if space == "" {
+			continue
+		}
+		err := tx.Query(ctx, insertConstraintSpaceStmt, dbConstraintSpace{
+			ConstraintUUID: constraintUUID,
+			Space:          space,
+		}).Run()
+		if internaldatabase.IsErrConstraintForeignKey(err) {
+			return errors.Errorf("inserting constraint space %q: %w", space, networkerrors.SpaceNotFound)
```

Comment:
I think we can do better with this error message. Something like errors.Errorf("setting model space constraints, constraint %q does not exist").Add(SpaceNotFound)